Add MASWE-0117 - Inadequate Permission Management (#3119)
OWASP/owasp-mastg: weaknesses/MASVS-PRIVACY/MASWE-0117.md

---
title: Inadequate Permission Management
id: MASWE-0117
alias: inadequate-permission-management
platform: ["android", "ios"]
profiles: ["P"]
mappings:
  masvs-v1: []
  masvs-v2: [MASVS-PRIVACY-1]
  cwe: [250]
refs:
- https://developer.apple.com/design/human-interface-guidelines/privacy#Requesting-permission
- https://developer.apple.com/documentation/uikit/protecting_the_user_s_privacy/requesting_access_to_protected_resources

> - https://developer.android.com/training/permissions/requesting#explain

Reviewer: Do we need these URLs with the anchor, as we anyway are referencing to the main page (https://developer.android.com/training/permissions/requesting)

- https://support.google.com/googleplay/android-developer/answer/9888170?hl=en
- https://developer.android.com/privacy-and-security/minimize-permission-requests
- https://developer.android.com/training/permissions/requesting

> - https://developer.android.com/training/permissions/requesting#remove-access

Reviewer: Do we need these URLs with the anchor, as we anyway are referencing to the main page (https://developer.android.com/training/permissions/requesting)

- https://developer.android.com/training/permissions/usage-notes
- https://arxiv.org/pdf/1905.02713
- https://arxiv.org/pdf/2203.10583

> - 

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

status: new
---

## Overview

> Inadequate permission management poses significant risks to user privacy and security in mobile apps. Permissions control access to sensitive device features, such as the camera, microphone, location, and storage, which are essential for app functionality. However, improper management, such as requesting excessive or unnecessary permissions, can lead to privacy violations, unauthorized data access, and erosion of user trust. For example, some apps retain access to sensitive resources long after they are no longer needed (even when platforms like Android 13 offer mechanisms to revoke unused permissions). Developers face the challenge of balancing functionality with privacy, as revoking permissions can disrupt app usability, forcing users to choose between privacy and functionality.

Reviewer: ```suggestion
Inadequate permission management poses significant risks to user privacy and security in mobile apps. Permissions control access to sensitive device features, such as the camera, microphone, location, and storage, which can be essential for app functionality. However, improper management, such as requesting excessive or unnecessary permissions, can lead to privacy violations, unauthorized data access, and erosion of user trust. For example, some apps retain access to sensitive resources long after they are no longer needed (even when platforms like Android 13 offer mechanisms to revoke unused permissions). Developers face the challenge of balancing functionality with privacy, as revoking permissions can disrupt app usability, forcing users to choose between privacy and functionality.
```

> Some apps, especially those that come pre-installed on devices, are often granted excessive privileges without needing explicit consent. Also, some regular apps may request permissions that are not necessary for their core functionality or request broader access than needed. For example, granting access to the camera may also grant access to the photo gallery, reducing user control. Privacy-friendly alternatives (e.g., coarse location or image picker) are often ignored.

Reviewer: ```suggestion
Some apps, especially those that come pre-installed on devices, are often granted excessive privileges without needing explicit consent.
```

This is redundant to what is mentioned in the modes of introduction. We could remove it from here.

Third-party libraries (SDKs) further complicate permission management by inheriting app permissions. The third-party services behind those libraries may continue to access data collected over the network even after permissions are revoked or the app is deleted. Users typically lack visibility and control over this data retention, which can then be used for marketing, profiling, or other purposes without the user's explicit consent or control.

## Modes of Introduction

- **Lack of Proactive Permission Revocation**: Not automatically revoking app permissions that are no longer necessary, resulting in unnecessary data access over time.  
- **Requesting Excessive Permissions**: Apps requesting more permissions than necessary for core functionality, often resulting in excessive data collection beyond what is required for the app to operate.
- **Lack of Use of Privacy-Friendly Alternatives**: Failing to use privacy-friendly alternatives to permissions that are less intrusive and provide users with more control over their data. For example, using coarse location instead of fine location, or using an image picker instead of requesting access to the camera and photo gallery.

## Impact

> - **Violation of User Privacy**: Users may have their personal data accessed unnecessarily leading to potential misuse of personal data, identity theft or surveillance.  

Reviewer: ```suggestion
- **Violation of User Privacy**: Users may have had their personal data unnecessarily accessed by mobile apps  leading to potential misuse of personal data, identity theft or surveillance.  
```

- **Loss of User Trust**: Users may lose trust in an app if it requests unnecessary permissions or does not allow them to revoke permissions that are no longer relevant. This can lead to negative reviews, lower user engagement, and reduced retention.
- **Legal and Compliance Issues**: Apps that improperly manage permissions may face non-compliance with privacy regulations like GDPR or CCPA, which require data minimization and appropriate user control over data access, resulting in potential fines, legal action, or removal from app stores.
- **Malicious Abuse:** Harmful apps can misuse permissions from privileged apps to record, track, or steal data without user consent.

> - **Security Breaches:** Once the collected sensitive data otherwise protected by permissions leaves the app, if it's poorly protected on the remote endpoints, it can be vulnerable to cyberattacks.

Reviewer: ```suggestion
- **Security Breaches:** Once the collected sensitive data otherwise protected by permissions leaves the app, it can be vulnerable to cyberattacks, if it's poorly protected on the remote endpoints.
```

- **Lack of Proactive Permission Revocation**: Not automatically revoking app permissions that are no longer necessary, resulting in unnecessary data access over time.
- **Inadequate Permission Explanations**: Failing to provide clear explanations for why each permission is required.

## Impact

- **Violation of User Privacy**: Users may have their personal data unnecessarily accessed by mobile apps, leading to potential misuse, identity theft, or surveillance.
- **Loss of User Trust**: Users may lose trust in an app if it requests unnecessary permissions or does not allow them to revoke permissions that are no longer relevant. This can lead to negative reviews, lower user engagement, and reduced retention.
- **Legal and Compliance Issues**: Apps that improperly manage permissions may face non-compliance with privacy regulations like GDPR or CCPA, which require data minimization and appropriate user control over data access, resulting in potential fines, legal action, or removal from app stores.
- **Malicious Abuse:** Harmful apps can misuse permissions from privileged apps to record, track, or steal data without user consent.

> - **Data Breaches:** Once sensitive data leaves the app, its safety can no longer be guaranteed and it may end up in large-scale data leaks.

Reviewer 2: Data Breaches: Once sensitive data leaves the app, its security can no longer be guaranteed, increasing the risk of large-scale data exposure.
```suggestion
- **Data Breaches:** Once sensitive data leaves the app, its security can no longer be guaranteed, increasing the risk of large-scale data exposure via data breaches.
```

> - **Data Breaches:** Once sensitive data leaves the app, it can be vulnerable to cyberattacks if it's not properlyprotected on remote endpoints, potentially leading to large-scale data leaks and reputational damage.

Author: ```suggestion
- **Data Breaches:** Once sensitive data leaves the app, it can be vulnerable to cyberattacks if it's not properly protected on remote endpoints, potentially leading to large-scale data leaks and reputational damage.
```

> - **Data Breaches:** Once sensitive data leaves the app, it can be vulnerable to cyberattacks if it's not properly protected on remote endpoints, potentially leading to large-scale data leaks and reputational damage.

Author: ```suggestion
- **Data Breaches:** Once sensitive data leaves the app, its safety can no longer be guaranteed and it may end up in large-scale data leaks.
```
